documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+Numerical requirements will need to be accessible in such cases so that lowering can select the right operations and/or propagate the requirements as flags on the operations.
+
+#### Scalar operands evaluation in elemental expressions
+##### Rule
+Scalar operands in elemental expressions must be evaluated only once.
+##### Rational
+10.1.10 point 2. says a scalar in an elemental expression must be treated as an array with _all elements equal_ to the scalar.
+This means that if the compiler cannot prove that re-evaluating such operands will yield the same result, it must be evaluated only once while evaluating the elemental expression. In other cases, a lowering that re-evaluate the scalar sub-expression for every element would be standard compliant. 
+However, in general, skipping useless evaluation should be a good thing regarding performance. So, rather than to do an analysis regarding whether or scalar operand must not be re-evaluated, it is easier and performance friendly to request that they should be evaluated only once.
+##### Example
+In `array + foo(x)` where `foo(x)` is scalar, foo will only be called once.
+##### Plans to enforce the rule while lowering to FIR
+The evaluation of scalar operands will be inserted before the loop evaluating the elemental expression. In the example above, the generated FIR should look like:
+```
+%1 = fir.call foo %x
+fir.loop %i 1 to n {
+  %2 = fir.extract_value %array %i 
+  %3 = fir.fadd  %2 %1
+  %4 = fir.coordinate_of %result %i
+  fir.store %3 to %4
+} 
+```
+#### Short-circuiting
+##### Rule
+Expression lowering to FIR must insert short-circuiting of operand evaluation that are guaranteed to not be performed in contexts defined in by f18 semantic extensions. Other short-circuiting opportunities should be left to later optimization phases.
+##### Rational
+If it needs to be guaranteed something will not be evaluated if unneeded, then that is semantic related, and it is unsafe to assume later FIR transformation and processing will respect this if short-circuiting was not explicitly inserted while lowering the expression.
+However, in general, the decision of whether to short-circuit an evaluation or not should be left to the optimizers based on the cost of the useless operation compared to the cost of the potential branch. Therefore, expression lowering should not add short-circuits that are not required because these cannot always be reversed in FIR (unless FIR transformations take the liberty to violate semantics which they probably should not).
+Note that it is unclear how well FIR transformation will be able to introduce short-circuits that are rather obvious looking at the `evaluate::Expr` tree, so this rule might get revisited.
+So far, the required short-circuits are:
+- `b.OR.SE`, `b.AND.SE` -> do not evaluate `SE` when `b` is false and `SE` has side effects.
```

Comment:
When `b` is true in the `.OR.` expr and `b` is false in the `.AND.` expr

Reply:
Thanks